### PR TITLE
[REVIEW] Enforce rule of 0 for memory resources

### PR DESCRIPTION
**Pull request by jrhemstad**
In line with what [Scott Meyer's says](http://scottmeyers.blogspot.com/2014/03/a-concern-about-rule-of-zero.html), makes Rule of 0 explicit by explicitly defaulting or deleting ctors/assignment operators as appropriate for all resources. 

I also cleaned up a few minor things I noticed while I was going through all the files, like cleaning up unnecessary includes.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: include/rmm/mr/device/fixed_multisize_memory_resource.hpp [hunk 1]
```diff
@@ -39,13 +39,10 @@
 using cudaStream_t = struct CUstream_st*;
 
 namespace rmm {
-
 namespace mr {
-
-namespace {
-
+namespace detail {
```
Comment (jrhemstad) [inline include/rmm/mr/device/fixed_multisize_memory_resource.hpp:43]: Anonymous namespace shouldn't be used in headers.

Comment (harrism) [reply include/rmm/mr/device/fixed_multisize_memory_resource.hpp:43]: TIL https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file

**Pull request merged**